File: source/native/src/SndParser.cpp
#include <Godot.hpp>
#include <File.hpp>
#include "SndParser.hpp"
#include "sff/data/ByteArray.hpp"
#include "sff/data/ByteArrayStream.hpp"
#include "sff/data/FileStream.hpp"
#include "snd/structs.h"

using namespace godot;

SndParser::SndParser()
{
    // pass
}

void SndParser::_init()
{
    // pass
}

Variant SndParser::get_sounds(String path)
{
    File* file = File::_new();
    Error error = file->open(path, File::ModeFlags::READ);
    Dictionary result;

    if (error != Error::OK) {
        Godot::print("Error opening snd file");
        return result;
    }

    SND_HEADER head;

    FileStream in(file);

    in >> head;

    if (strcmp(&head.signature[0], "ElecbyteSnd") != 0) {
        Godot::print("Invalid snd signature");
        return result;
    }

    file->seek(head.subheader_offset);

    for (int i = 0; i < 4096; i++) {
        SND_SUBHEADER subheader;
        in >> subheader;

        if (subheader.length == 0) {
            break;
        }

        ByteArray tmpArr;
        in.readRawData(tmpArr, subheader.length);

        ByteArrayStream tmpArrayStream = ByteArrayStream(tmpArr);
        WAV_HEADER wavHeader;
        tmpArrayStream >> wavHeader;

        Dictionary dict;
        dict["groupno"] = subheader.groupno;
        dict["soundno"] = subheader.soundno;
        dict["audio_format"] = wavHeader.audioFormat;
        dict["num_channels"] = wavHeader.numChannels;
        dict["sample_rate"] = wavHeader.sampleRate;
        dict["byte_rate"] = wavHeader.byteRate;
        dict["block_align"] = wavHeader.blockAlign;
        dict["bits_per_sample"] = wavHeader.bitsPerSample;
        dict["data"] = tmpArr.toSigned();

        std::string key = std::to_string(subheader.groupno) + "-" + std::to_string(subheader.soundno);
        result[key.c_str()] = dict;

        tmpArr.clear();

        if (subheader.next > 0 && subheader.next < file->get_len()) {
            file->seek(subheader.next);
        } else {
            break;
        }
    }

    file->close();

    return result;
}

void SndParser::_register_methods() {
    register_method("get_sounds", &SndParser::get_sounds);
}

File: source/native/src/snd/snd_parser.rs
use crate::sff::data::{BufferReader, DataError, DataReader, FileReader};
use crate::snd::structs::{FileHeader, SubHeader, WavHeader};
use gdnative::api::File;
use gdnative::api::AudioStreamSample;
use gdnative::prelude::*;

#[derive(NativeClass)]
#[inherit(Reference)]
pub struct SndParser {}

#[methods]
impl SndParser {
    pub fn new(_owner: &Reference) -> Self {
        SndParser {}
    }

    #[method]
    pub fn read_sounds(&self, path: String) -> Variant {
        let result = self.read_file(path);

        if let Ok(dict) = result {
            return dict.to_variant();
        } else if let Err(message) = result {
            godot_error!("{}", message);
        }

        Variant::nil()
    }

    pub fn read_file(&self, path: String) -> Result<Dictionary, DataError> {
        let file = File::new();
        let open_result = file.open(path, File::READ);

        if let Err(detail) = open_result {
            return Result::Err(DataError::new(format!(
                "Error opening snd file: {}",
                detail
            )));
        }

        let mut reader = FileReader::new(&file);
        let head = FileHeader::read(&mut reader);

        if head.signature != "ElecbyteSnd" {
            file.close();
            return Result::Err(DataError::new(format!(
                "Snd invalid signature: {}",
                head.signature
            )));
        }

        file.seek(head.subheader_offset as i64);

        let result = Dictionary::new();

        for _ in 0..4096 {
            if reader.eof() {
                break;
            }

            let subheader = SubHeader::read(&mut reader);

            if subheader.length == 0 {
                break;
            }

            let tmp_arr = reader.get_buffer(subheader.length as usize);
            let mut tmp_arr_reader = BufferReader::new(&tmp_arr);
            let wav_header = WavHeader::read(&mut tmp_arr_reader);

            let stream = AudioStreamSample::new();
            stream.set_data(to_signed(&tmp_arr));
            stream.set_mix_rate(wav_header.sample_rate as i64);
            stream.set_stereo(wav_header.num_channels == 2);
            match wav_header.bits_per_sample {
                8 => stream.set_format(AudioStreamSample::FORMAT_8_BITS),
                16 => stream.set_format(AudioStreamSample::FORMAT_16_BITS),
                _ => {
                    godot_warn!("invalid bits_per_sample: {}", wav_header.bits_per_sample);
                }
            };

            let key = format!("{}-{}", subheader.groupno, subheader.soundno);
            result.insert(key, stream);

            if subheader.next > 0 && subheader.next < file.get_len() as u32 {
                file.seek(subheader.next as i64);
            } else {
                break;
            }
        }

        file.close();

        Result::Ok(result.into_shared())
    }
}

fn to_signed(source: &[u8]) -> PoolArray<u8> {
    source
        .iter()
        .map(|value| ((*value as i16) - 128) as u8)
        .collect()
}
